### PR TITLE
Add syncronization for Gas Price Database and On Chain Database on startup

### PR DESCRIPTION
**Pull request by MitchTurner**
Closes https://github.com/FuelLabs/fuel-core/issues/2030

In the unlikely situation that the metadata db is behind the on chain database, we should apply the values from each block to the algorithm updater to produce reasonable values.

For now, we should be able to use the updater deterministically, but this might become more difficult as we include the DA values in the updater.

We also define the behavior for when the gas price db is beyond the on-chain db. For now, we are assuming that the blocks produced with the gas prices in the db were either reverted or not properly committed to the on_chain db, and we just revert the gas price db to the height of the on chain db. i.e. **the on chain db is the source of truth.**

## Checklist
- [ ] Breaking changes are clearly marked as such in the PR description and changelog
- [ ] New behavior is reflected in tests
- [ ] [The specification](https://github.com/FuelLabs/fuel-specs/) matches the implemented behavior (link update PR if changes are needed)

### Before requesting review
- [ ] I have reviewed the code myself
- [ ] I have created follow-up issues caused by this PR and linked them here

### After merging, notify other teams

[Add or remove entries as needed]

- [ ] [Rust SDK](https://github.com/FuelLabs/fuels-rs/)
- [ ] [Sway compiler](https://github.com/FuelLabs/sway/)
- [ ] [Platform documentation](https://github.com/FuelLabs/devrel-requests/issues/new?assignees=&labels=new+request&projects=&template=NEW-REQUEST.yml&title=%5BRequest%5D%3A+) (for out-of-organization contributors, the person merging the PR will do this)
- [ ] Someone else?

### CODE REVIEW

File: crates/fuel-core/src/service/sub_services.rs [hunk 1]
```diff
@@ -337,3 +372,97 @@ pub fn init_sub_services(
...
+        l2_block_fullness_threshold_percent: config.gas_price_threshold_percent,
+    });
+    if metadata_height < latest_block_height {
+        let mut metadata_storage = StructuredStorage::new(gas_price_db);
+        let metadata = metadata_storage
+            .get_metadata(&metadata_height.into())?
+            .unwrap_or(default_metadata);
+        let mut inner: AlgorithmUpdater = metadata.into();
+        let AlgorithmUpdater::V0(ref mut updater) = &mut inner;
+        for height in (metadata_height + 1)..=latest_block_height {
```
Comment (MitchTurner) [inline crates/fuel-core/src/service/sub_services.rs:414]: This iterates over all the blocks and applies them to the algorithm. This should work for now, but could become confusing when we include the da price calculation in the future.

The confusing part is that the blocks already include the historical gas prices, so do we really need to re-calculate them via the updater?

File: crates/fuel-core/src/service/sub_services/algorithm_updater.rs [hunk 2]
```diff
@@ -0,0 +1,209 @@
...
+                )?;
+                metadata_storage
+            }
+            Ordering::Greater => {
+                revert_gas_price_db_to_height(
+                    &mut gas_price_db,
+                    latest_block_height.into(),
+                )?;
+                StructuredStorage::new(gas_price_db)
+            }
```
Comment (MitchTurner) [inline crates/fuel-core/src/service/sub_services/algorithm_updater.rs:107]: I don't like how complex and special-casey this code is. A lot of the complexity is coming from the fact that we don't store the metadata for the genesis block, so we have to manually pass the `genesis_metadata` around for certain cases.

I also think calling `revert_gas_price_db_to_height` in two different locations is a smell, so I might sleep on this code and see if a clear approach arises. Kinda been staring at it too long to think straight.

**Comment by MitchTurner** [comment]
We can create an `InitializationTask` similar to the `graph_ql::Task` that will get run in a non-blocking thread on startup of the service.

This will help us contain the logic for advancing the gas price to the service and just use the OnChain DB to create a list of blocks that still need to be added.

**Review by Dentosal — DISMISSED** [review]
Nice work with the tidy-up commit.

**Review by Dentosal — DISMISSED** [review]

**Review by Dentosal — APPROVED** [review]

**Review by Voxelot — APPROVED** [review]

**Pull request merged**